src/resolve/resolved-mdns-browse-services.c

```diff
@@ -0,0 +1,454 @@
+
+        if (isempty(name))
+                name = NULL;
+        else if (!dns_service_name_is_valid(name))
+                return varlink_error_invalid_parameter(link, JSON_VARIANT_STRING_CONST("name"));
+
+        if (isempty(type))
+                type = NULL;
+        else if (!dnssd_srv_type_is_valid(type))
+                return varlink_error_invalid_parameter(link, JSON_VARIANT_STRING_CONST("type"));
+
+        r = dns_name_is_valid(domain);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return varlink_error_invalid_parameter(link, JSON_VARIANT_STRING_CONST("domain"));
+
+        r = dns_question_new_browse_service(&question_utf8, name, type, domain, false);
+        if (r < 0)
+                return r;
+
+        r = dns_question_new_browse_service(&question_idna, name, type, domain, true);
+        if (r < 0)
+                return r;
+
+        ss = new(mDnsServiceSubscriber, 1);
+        if (!ss)
+                return log_oom();
+
+        *ss = (mDnsServiceSubscriber) {
+                .n_ref = 1,
+                .m = m,
+                .delay = 0,
+                .link = varlink_ref(link),
+                .question_utf8 = dns_question_ref(question_utf8),
+                .question_idna = dns_question_ref(question_idna),
+                .ifindex = ifindex,
+                .token = token,
+                .flags = SD_RESOLVED_NO_CACHE | SD_RESOLVED_MDNS,
+        };
+
+        ss->key = dns_question_first_key(ss->question_utf8);
+
+        r = sd_event_add_time(m->event,
+                        &ss->schedule_event,
+                        CLOCK_BOOTTIME,
+                        usec_add(now(CLOCK_BOOTTIME), (ss->delay * USEC_PER_SEC)),
+                        0,
+                        mdns_next_query_schedule,
+                        ss);
+
+        m->dns_service_subscriber = TAKE_PTR(ss);
+
+        return 0;
+}
+
+mDnsServiceSubscriber *mdns_service_subscriber_free(mDnsServiceSubscriber *ss) {
+        DnsQuery *q;
+
+        assert(ss);
+
+        if (ss->schedule_event)
+                sd_event_source_disable_unref(ss->schedule_event);
+
+
+        q = varlink_get_userdata(ss->link);
+        if (q) {
+                if (DNS_TRANSACTION_IS_LIVE(q->state))
+                        dns_query_complete(q, DNS_TRANSACTION_ABORTED);
+        }
+
+        LIST_FOREACH(mdns_services, service, ss->mdns_services)
+                mdns_remove_service(ss, service);
+
+        dns_question_unref(ss->question_idna);
+        dns_question_unref(ss->question_utf8);
+
+        if (ss->link)
+                varlink_unref(ss->link);
+
+        return mfree(ss);
+}
+
+mDnsGoodbyeParams* mdns_goodbye_params_free(mDnsGoodbyeParams *p) {
+        if (!p)
+                return NULL;
+
+        p->ss = mdns_service_subscriber_unref(p->ss);
+        return mfree(p);
+}
+
+DEFINE_TRIVIAL_REF_UNREF_FUNC(mDnsServiceSubscriber, mdns_service_subscriber, mdns_service_subscriber_free);
```

Comment:
Please move this soon after the declaration of `mdns_service_subscriber_free()`.